File: C_TRR/3.c
// Viết bằng C
// Bài 3: Sinh tập con của một tập hợp.

#include <stdio.h>

void generateSubsets(int arr[], int n) 
{
    int subsetCount = 1 << n; // 2^n
    for (int i = 0; i < subsetCount; i++) 
    {
        printf("{ ");
        for (int j = 0; j < n; j++) 
        {
            if (i & (1 << j)) // Kiểm tra bit j
                printf("%d ", arr[j]);
        }
        printf("}\n");
    }
}

int main() 
{
    int arr[] = {1, 2, 3};
    int n = 3;
    printf("Tập con:\n");
    generateSubsets(arr, n);
    return 0;
}

File: Python_TRR/3.py
# Viết bằng Python
# Bài 3: Sinh tập con của một tập hợp.

# Giải thích: Sinh tập con là liệt kê tất cả các tập hợp con có thể từ một tập hợp cho trước.

from itertools import combinations # Sử dụng hàm combinations từ thư viện itertools.

def generate_subsets(arr): # Hàm sinh tập con. 
    n = len(arr) # Số phần tử của tập hợp.
    subsets = [] # Danh sách chứa tất cả các tập con.
    for r in range(n + 1): # Duyệt qua tất cả các số phần tử của tập con.
        subsets.extend(combinations(arr, r))
    return subsets

# Ví dụ
arr = [1, 2, 3] # Tập hợp ban đầu.
print("Tập con:") # In ra tập
for subset in generate_subsets(arr):
    print(subset)

# Kết quả
# Tập con:
# ()    
# (1,)
# (2,)
# (3,)
# (1, 2)
# (1, 3)
# (2, 3)
# (1, 2, 3)